kratos/solving_strategies/convergencecriterias/residual_criteria.h

```diff
@@ -210,19 +206,48 @@ class ResidualCriteria
      * @brief This function initializes the solution step
      * @param rModelPart Reference to the ModelPart containing the problem.
      * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
-     * @param A System matrix (unused)
-     * @param Dx Vector of results (variations on nodal variables)
-     * @param b RHS vector (residual + reactions)
+     * @param rA System matrix (unused)
+     * @param rDx Vector of results (variations on nodal variables)
+     * @param rb RHS vector (residual + reactions)
      */
     void InitializeSolutionStep(
         ModelPart& rModelPart,
         DofsArrayType& rDofSet,
-        const TSystemMatrixType& A,
-        const TSystemVectorType& Dx,
-        const TSystemVectorType& b
+        const TSystemMatrixType& rA,
+        const TSystemVectorType& rDx,
+        const TSystemVectorType& rb
         ) override
     {
-        mInitialResidualIsSet = false;
+        BaseType::InitializeSolutionStep(rModelPart, rDofSet, rA, rDx, rb);
+
+        mActiveDofs.resize(rDofSet.size());
+
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(mActiveDofs.size()); ++i)
+            mActiveDofs[i] = true;
+
+        #pragma omp parallel for 
+        for (int i = 0; i < static_cast<int>(rDofSet.size()); i++) {
+            auto it_dof = rDofSet.begin() + i;
+            if(it_dof->IsFixed())
+                mActiveDofs[it_dof->EquationId()] = false;
+        }
+        
+        for(auto& mpc : rModelPart.MasterSlaveConstraints())
+        {
+            auto& slave_dofs = mpc.GetSlaveDofsVector();
+            auto& master_dofs = mpc.GetMasterDofsVector();
+
+            for(auto& dof : slave_dofs)
+                mActiveDofs[dof->EquationId()] = false;
+            for(auto& dof : master_dofs)
```

Comment:
Why do you exclude the master dofs?

Or why excluding the mpcs at all?

Reply:
the residual on the dirichlet dofs and on the mpcs is non zero (it converges to the reactions), what goes to zero is only the norm of the residual on the non dirichlet (and nonmaster/slave nodes)

Reply:
ok

FYI @adityaghantasala